mmengine/visualization/visualizer.py

```diff
@@ -194,12 +188,30 @@ def __init__(
             if None not in names and len(set(names)) != len(names):
                 raise RuntimeError('The name fields cannot be the same')
 
-            save_dir = osp.join(save_dir, 'vis_data')
+            # check if vis_backends are initialized
+            if save_dir is None:
+                for vis_backend in vis_backends:
+                    if vis_backend.get('save_dir') is None:
+                        print_log(
+                            '`Visualizer` backend is not initialized '
+                            'because save_dir is None '
+                            'and at least one vis_backend has no save_dir.',
+                            logger='current',
+                            level=logging.WARNING)
+                        break
 
-            for vis_backend in vis_backends:
-                name = vis_backend.pop('name', vis_backend['type'])
-                vis_backend.setdefault('save_dir', save_dir)
-                self._vis_backends[name] = VISBACKENDS.build(vis_backend)
+            else:
+                save_dir = osp.join(save_dir, 'vis_data')
+                for vis_backend in vis_backends:
+                    name = vis_backend.pop('name', vis_backend['type'])
+                    vis_backend.setdefault('save_dir', save_dir)
+                    self._vis_backends[name] = VISBACKENDS.build(vis_backend)
```

Comment:
```python
# check if vis_backends are initialized
if save_dir is None:
    for vis_backend in vis_backends:
        if vis_backend.get('save_dir') is None:
            print_log(
                '`Visualizer` backend is not initialized '
                'because save_dir is None '
                'and at least one vis_backend has no save_dir.',
                logger='current',
                level=logging.WARNING)
            break

else:
    save_dir = osp.join(save_dir, 'vis_data')
    for vis_backend in vis_backends:
        name = vis_backend.pop('name', vis_backend['type'])
        vis_backend.setdefault('save_dir', save_dir)
        self._vis_backends[name] = VISBACKENDS.build(vis_backend)
```
->
```python
if save_dir is not None:
    save_dir = osp.join(save_dir, 'vis_data')
for vis_backend in vis_backends:
    name = vis_backend.pop('name', vis_backend['type'])
    if 'save_dir' not in vis_backend:
        if save_dir is not None:
            vis_backend[save_dir'] = save_dir
        else:
            raise ValueError(
                f'save_dir should be specified in {vis_backend} or '
                f'{self.__class__.__name__}')
    self._vis_backends[name] = VISBACKENDS.build(vis_backend)
```

Reply:
Failed in test case https://github.com/open-mmlab/mmengine/blob/main/tests/test_visualizer/test_visualizer.py#L77-L79.
I think test_visualizer.py should be modified simultaneously.